plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java

```diff
@@ -0,0 +1,201 @@
+    Assert.assertTrue(webInfFolders.isEmpty());
+  }
+
+  @Test
+  public void testFindAllWebInfFolders() throws CoreException {
+    IProject project = projectCreator.getProject();
+    createPath(project, new Path("src/my-webapp/WEB-INF"));
+
+    List<IFolder> webInfFolders =
+        FacetUtil.findAllWebInfFolders(project);
+    Assert.assertEquals(1, webInfFolders.size());
+    Assert.assertEquals(project.getFolder("src/my-webapp/WEB-INF"), webInfFolders.get(0));
+  }
+
+  @Test
+  public void testFindAllWebInfFolders_multipleFolders() throws CoreException {
+    IProject project = projectCreator.getProject();
+    createPath(project, new Path("webapps/first-webapp/WEB-INF"));
+    createPath(project, new Path("webapps/second-webapp/WEB-INF"));
+    createPath(project, new Path("third-webapp/WEB-INF"));
+    createPath(project, new Path("WEB-INF"));
+
+    List<IFolder> webInfFolders = FacetUtil.findAllWebInfFolders(project);
+    Assert.assertEquals(4, webInfFolders.size());
+    Assert.assertTrue(webInfFolders.contains(project.getFolder("webapps/first-webapp/WEB-INF")));
+    Assert.assertTrue(webInfFolders.contains(project.getFolder("webapps/second-webapp/WEB-INF")));
+    Assert.assertTrue(webInfFolders.contains(project.getFolder("third-webapp/WEB-INF")));
+    Assert.assertTrue(webInfFolders.contains(project.getFolder("WEB-INF")));
+  }
+
+  @Test
+  public void testFindMainWebAppDirectory_noWebInfFolders() {
+    IPath mainWebApp = FacetUtil.findMainWebAppDirectory(projectCreator.getProject());
+    Assert.assertNull(mainWebApp);
+  }
+
+  @Test
+  public void testFindMainWebAppDirectory() throws CoreException {
+    createPath(projectCreator.getProject(), new Path("webapps/first-webapp/WEB-INF"));
+    IPath mainWebApp = FacetUtil.findMainWebAppDirectory(projectCreator.getProject());
+    Assert.assertEquals(new Path("webapps/first-webapp"), mainWebApp);
+  }
+
+  @Test
+  public void testFindMainWebAppDirectory_returnsFolderWithWebXml() throws CoreException {
+    IProject project = projectCreator.getProject();
+    createPath(project, new Path("webapps/first-webapp/WEB-INF"));
+    createEmptyFile(project, new Path("webapps/second-webapp/WEB-INF/web.xml"));
+    createPath(project, new Path("third-webapp/WEB-INF"));
+    createPath(project, new Path("WEB-INF"));
+
+    IPath mainWebApp = FacetUtil.findMainWebAppDirectory(project);
+    Assert.assertEquals(new Path("webapps/second-webapp"), mainWebApp);
+  }
+
+  private static void createEmptyFile(IProject project, IPath relativePath) throws CoreException {
+    createPath(project, relativePath.removeLastSegments(1));
+
+    InputStream emptyStream = new ByteArrayInputStream(new byte[0]);
+    project.getFile(relativePath).create(emptyStream, false /* force */, null /* monitor */);
+    Assert.assertTrue(project.getFile(relativePath).exists());
+  }
+
+  private static void createPath(IContainer parent, IPath relativePath) throws CoreException {
```

Comment:
Add a reference to https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1573 and add this method to the issue's description so we can keep track of it.